jax_md/energy.py

```diff
@@ -891,6 +893,285 @@ def compute_fn(R, neighbor, **kwargs):
+      i += 1
+      line_add = raw_text[i]
+      if not line_add:
+        continue
+      
+      line += ' ' + line_add
+      words = line.strip().split()
+      nwords = len(words)
+      i +=1
+
+    if nwords != params_per_line:
+      raise ValueError('Incorrect format: %d not in %d'
+        % (nwords, params_per_line))
+
+    words[3:] = f32(words[3:])
+    params_tmp = {'element1': words[0], 
+                  'element2': words[1],
+                  'element3': words[2],
+                  'mTf': words[3], 'gamma': words[4], 'lam3': words[5], 
+                  'cTf': words[6], 'dTf': words[7], 'hTf': words[8],
+                  'nTf': words[9], 'beta': words[10], 
+                  'lam2': words[11], 'B': words[12],
+                  'R': words[13], 'D': words[14],
+                  'lam1': words[15], 'A': words[16]}
+    params.append(params_tmp)
+
+  return params
+
+def _ters_cutoff(dr, R, D) -> Array:
+  """The cut-off function of the Tersoff potential.
+  Args: 
+    R: A scalar that is the average of inner and outer cutoff radii
+    D: A scalar that is the half of the difference 
+       between inner and outer cutoff radii
+
+  Returns:
+    cut-off values
+  """
+  outer = jnp.where(dr < R+D,
+                    0.5*(1.0 - jnp.sin(jnp.pi/2*(dr - R)/D)),
+                    0)    
+  inner = jnp.where(dr < R-D, 1, outer)
+  return inner
+
+def _ters_bij(dRij, dRik, neighbor, mask_ijk,
+              R, D, c, d, h, lam3, beta, n, m) -> Array:
+  """The bond-order term of the Tersoff potential.
+  Args: 
+    R: A scalar that is the average of inner and outer cutoff radii
+    D: A scalar that is the half of the difference 
+       between inner and outer cutoff radii
+    c, d, h: scalars that determines angle penalty value in bond-order 
+    Lam3, m: scalars that determines distance penalty value in bond-order
+    beta, n: scalars that determines bond-order
+
+  Returns:
+    bond-order values between i and j atoms
+  """
+  drij = space.distance(dRij)
+  drik = space.distance(dRik)
+  
+  # compute g_ijk - angle penalty value
+  costheta = quantity.cosine_angles(dRij)
+  gijk = 1.0 + (c**2 / d**2) - (c**2 / (d**2 + (h - costheta)**2))
+  gijk *= mask_ijk
+  gijk = vmap(lambda x: jax.ops.index_update(x, jnp.diag_indices(x.shape[0]), 0))(gijk)
+
+  # compute exponential term - distance penalty value
+  # todo: it can be converted by space.map_neighbor
+  # todo: masking procedure should be reduced
+  dr_diff = vmap(vmap(jnp.subtract, (None, 0)))(drij, drik)
+  explr3 = jnp.exp(lam3**m * dr_diff**m)
+  explr3 *= mask_ijk
+  explr3 = vmap(lambda x: jax.ops.index_update(x, jnp.diag_indices(x.shape[0]), 0))(explr3)
+
+  # compute fC with dr_ik
+  fC = _ters_cutoff(drik, R, D)
+
+  # compute zeta without diagonal term
+  tmp = jnp.multiply(gijk, explr3)
+  zeta_ij = jnp.einsum('ik,ijk->ij', fC, tmp)
+  bij = (1 + (beta*zeta_ij)**n)**(-1 / 2 / n)
+  
+  return bij
+
+def _ters_attractive(B: f32, lam2: f32, 
```

Comment:
Can we keep (close to) the same argument order between `ters_attractive` and `ters_bij`? E.g. can we keep the `dR` and `mask` either at the beginning or the end?

Reply:
Yes. It is much easier when someone reads the code. Thanks very much.